server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java

```diff
@@ -90,20 +91,37 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
                     throw new IllegalStateException("cannot add duplicate task: " + a);
                 }, IdentityHashMap::new));
             LinkedHashSet<BatchedTask> newTasks = new LinkedHashSet<>(tasks);
-            tasksPerBatchingKey.merge(firstTask.batchingKey, newTasks, (existingTasks, updatedTasks) -> {
-                for (BatchedTask existing : existingTasks) {
-                    // check that there won't be two tasks with the same identity for the same batching key
-                    BatchedTask duplicateTask = tasksIdentity.get(existing.getTask());
-                    if (duplicateTask != null) {
-                        throw new IllegalStateException(
-                            "task ["
-                                + duplicateTask.describeTasks(Collections.singletonList(existing))
-                                + "] with source ["
-                                + duplicateTask.source
-                                + "] is already queued"
-                        );
+            // Need to maintain below order in which task identity map and task map are updated.
+            // For insert: First insert identity in taskIdentity map with dup check and then insert task in taskMap.
+            // For remove: First remove task from taskMap and then remove identity from taskIdentity map.
+            // We are inserting identity first and removing at last to ensure no duplicate tasks are enqueued.
+            // Changing this order might lead to duplicate tasks in queue.
+            taskIdentityPerBatchingKey.merge(firstTask.batchingKey, tasksIdentity, (existingIdentity, updatedIdentity) -> {
+                Map<Object, BatchedTask> existingIdentities = taskIdentityPerBatchingKey.computeIfAbsent(
+                    firstTask.batchingKey,
+                    k -> new IdentityHashMap<>()
+                );
```

Comment:
Why do we need computeIfAbsent here ? automatically merge will provide existing value as first argument and new value (passed in 2nd argument of merge) as 2nd argument of remapping function. 

So can't we use that information rather than again doing a computeIfAbsent ? - anyway internal remapping function will never be executed, if key is not present. It is only executed if key is present and we want to modify it.

Reply:
Also not sure if we should again take the bucket level lock once lock has already been taken as part of merge method execution.